core/src/main/java/com/spotify/ffwd/domain/SearchDomainDiscovery.kt

```diff
@@ -0,0 +1,120 @@
+        Type(SearchDomainDiscovery.Static::class),
+        Type(SearchDomainDiscovery.DynamicGcpMetadataMapping::class))
+interface SearchDomainDiscovery {
+    /**
+     * Discovers the search domain, which is used for domain resolution
+     */
+    fun discover(): Optional<String>
+
+    class Empty : SearchDomainDiscovery {
+
+        override fun discover(): Optional<String> {
+            return Optional.empty()
+        }
+    }
+
+    @JsonTypeName("static")
+    data class Static @java.beans.ConstructorProperties("domain")
+    constructor(val domain: String) : SearchDomainDiscovery {
+
+        override fun discover(): Optional<String> {
+            return Optional.of(domain)
+        }
+    }
+
+    @JsonTypeName("dynamic-gcp-metadata-mapping")
+    data class DynamicGcpMetadataMapping @java.beans.ConstructorProperties("mapping")
+    constructor(val mapping: Map<String, String>) : SearchDomainDiscovery {
+
+        override fun discover(): Optional<String> {
+            val content = doMetadataRequest()
+            val zone = parseZone(content)
+            val domain = mapping[zone]
+                    ?: throw RuntimeException("No domain mapping configured for zone: $zone")
+
+            return Optional.of(domain)
+        }
+
+        /**
+         * Parse zone out of metadata response. Example of content:
+         * projects/12345678/zones/europe-west1-d
+         */
+        private fun parseZone(content: String): String {
+            val trimmedContent = content.trim { it <= ' ' }
+            val parts = trimmedContent.split("/".toRegex()).dropLastWhile({ it.isEmpty() }).toTypedArray()
+
+            if (parts.size != 4) {
+                throw RuntimeException(
+                        "Metadata request returned unexpected content: $content")
+            }
+
+            return parts[3]
+        }
+
+        private fun doMetadataRequest(): String {
+            // TODO: Add a timeout to the metadata call.
+            try {
+                val requestUrl = URL(
+                        "http://metadata.google.internal/computeMetadata/v1" + "/instance/zone")
+                val connection = requestUrl.openConnection() as HttpURLConnection
+                connection.setRequestProperty("Metadata-Flavor", "Google")
+                val responseCode = connection.responseCode
+
+                if (responseCode != 200) {
+                    throw RuntimeException(
+                            "Metadata request returned unexpected code: $responseCode")
+                }
+
+                return CharStreams.toString(
+                        InputStreamReader(connection.inputStream, StandardCharsets.UTF_8))
+            } catch (e: Exception) {
+                throw RuntimeException(e)
+            }
+
+        }
+    }
+
+    companion object {
+        /**
+         * Default implementation for http discovery when nothing else is configured.
+         */
+        fun supplyDefault(): SearchDomainDiscovery {
```

Comment:
You can add the annotation `@JvmStatic` and access this the same as the Java static function, `SearchDomainDiscovery::supplyDefault`, instead of having to reference `.Companion`.